README.md
# unused-files-webpack-plugin
> Glob all files that are not compiled by webpack under webpack's context

[![Version][npm-image]][npm-url] [![Travis CI][travis-image]][travis-url]

## Installation

```bash
npm i --save-dev unused-files-webpack-plugin
```

## Requirements

* Node `^8.9.4`
* Webpack `^4.0.0`


## Usage

### `webpack.config.js`

```js
const { UnusedFilesWebpackPlugin } = require("unused-files-webpack-plugin");

module.exports = {
  plugins: [
    new UnusedFilesWebpackPlugin(options),
  ],
};
```


## Options

```js
new UnusedFilesWebpackPlugin(options)
```

### options.patterns

The (array of) pattern(s) to glob all files within the context.

* Default: `["**/*.*"]`
* Directly pass to [`glob-all(patterns)`](https://github.com/jpillora/node-glob-all#api)

### options.failOnUnused

Emit _error_ instead of _warning_ in webpack compilation result.

* Default: `false`
* Explicitly set it to `true` to enable this feature

### options.globOptions

The options object pass to second parameter of `glob-all`.

* Default: `{ignore: "node_modules/**/*"}`
* Directly pass to [`glob-all(pattern, globOptions)`](https://github.com/jpillora/node-glob-all#api), which then pass to [`glob(…, globOptions)`](https://github.com/isaacs/node-glob#options)

#### globOptions.ignore

Ignore pattern for glob. Can be a String or an Array of String.

* Default: `"node_modules/**/*"`
* Pass to: [`options.ignore`](https://github.com/isaacs/node-glob#options)

#### globOptions.cwd

Current working directory for glob. If you don't set explicitly, it defaults to the `context` specified by your webpack compiler at runtime.

* Default: `webpackCompiler.context`
* Pass to: [`options.cwd`](https://github.com/isaacs/node-glob#options)
* See also: [`context` in webpack](https://webpack.js.org/configuration/entry-context/#context)


## Contributing

[![devDependency Status][david-dm-image]][david-dm-url]

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


[npm-image]: https://img.shields.io/npm/v/unused-files-webpack-plugin.svg?style=flat-square
[npm-url]: https://www.npmjs.org/package/unused-files-webpack-plugin
[travis-image]: https://img.shields.io/travis/tomchentw/unused-files-webpack-plugin.svg?style=flat-square
[travis-url]: https://travis-ci.org/tomchentw/unused-files-webpack-plugin

[david-dm-image]: https://img.shields.io/david/dev/tomchentw/unused-files-webpack-plugin.svg?style=flat-square
[david-dm-url]: https://david-dm.org/tomchentw/unused-files-webpack-plugin#info=devDependencies



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="4.0.0"></a>
# [4.0.0](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v3.4.0...v4.0.0) (2021-07-10)


### Features

* **index.js:** Use webpack 5 hook API. ([c4d320e](https://github.com/tomchentw/unused-files-webpack-plugin/commit/c4d320e))



<a name="3.4.0"></a>
# [3.4.0](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v3.3.0...v3.4.0) (2018-03-14)


### Features

* **package.json:** upgrade dependencies and devDependencies ([250a6cb](https://github.com/tomchentw/unused-files-webpack-plugin/commit/250a6cb))



<a name="3.3.0"></a>
# [3.3.0](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v3.2.0...v3.3.0) (2018-03-14)


### Features

* add webpack v4 support ([#21](https://github.com/tomchentw/unused-files-webpack-plugin/issues/21)) ([bde0c36](https://github.com/tomchentw/unused-files-webpack-plugin/commit/bde0c36)), closes [#20](https://github.com/tomchentw/unused-files-webpack-plugin/issues/20)



<a name="3.2.0"></a>
# [3.2.0](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v3.1.0...v3.2.0) (2017-10-23)


### Features

* **index.js:** support multiple patterns ([961b1d1](https://github.com/tomchentw/unused-files-webpack-plugin/commit/961b1d1)), closes [#2](https://github.com/tomchentw/unused-files-webpack-plugin/issues/2)
* **package.json:** add `glob-all` and `warning` to dependencies ([3961606](https://github.com/tomchentw/unused-files-webpack-plugin/commit/3961606))
* **package.json:** remove `glob` from dependencies ([03e08b6](https://github.com/tomchentw/unused-files-webpack-plugin/commit/03e08b6))



<a name="3.1.0"></a>
# [3.1.0](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v3.0.2...v3.1.0) (2017-10-21)


### Features

* **index.js:** rewrite with async function ([0754e72](https://github.com/tomchentw/unused-files-webpack-plugin/commit/0754e72))
* **package.json:** add `babel-runtime`, `semver` and `util.promisify` to dependencies ([28cf6c5](https://github.com/tomchentw/unused-files-webpack-plugin/commit/28cf6c5))



<a name="3.0.2"></a>
## [3.0.2](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v3.0.1...v3.0.2) (2017-06-13)


### Bug Fixes

* **index.js:** avoid creating a new object in reduce loop ([8dcf206](https://github.com/tomchentw/unused-files-webpack-plugin/commit/8dcf206))



<a name="3.0.1"></a>
## [3.0.1](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v3.0.0...v3.0.1) (2017-02-13)


### Bug Fixes

* **index.js:** prevent freeze when using dev server without bail ([#12](https://github.com/tomchentw/unused-files-webpack-plugin/issues/12)) ([b397e8d](https://github.com/tomchentw/unused-files-webpack-plugin/commit/b397e8d))



<a name="3.0.0"></a>
# [3.0.0](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v2.0.5...v3.0.0) (2016-09-22)


### Bug Fixes

* **CHANGELOG.md:** deprecated v2.0.5 ([b7dfd07](https://github.com/tomchentw/unused-files-webpack-plugin/commit/b7dfd07))
* **index.js:** if bail is set, callback with error ([14b510a](https://github.com/tomchentw/unused-files-webpack-plugin/commit/14b510a))


### BREAKING CHANGES

* index.js: error reporting behaviour changed when bail is set

Before: if bail is set, there's still an error from UnusedFilesWebpackPlugin

After: if bail is set, no error will be emitted by UnusedFilesWebpackPlugin



<a name="2.0.5"></a>
## (Deprecated) [2.0.5](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v2.0.4...v2.0.5) (2016-09-20)

*Deprecated* due to https://github.com/tomchentw/unused-files-webpack-plugin/pull/11#issuecomment-248652421

### Bug Fixes

* **index.js:** if bail is set, callback with error ([#11](https://github.com/tomchentw/unused-files-webpack-plugin/issues/11)) ([23b85ad](https://github.com/tomchentw/unused-files-webpack-plugin/commit/23b85ad)), closes [#10](https://github.com/tomchentw/unused-files-webpack-plugin/issues/10)



<a name="2.0.4"></a>
## [2.0.4](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v2.0.3...v2.0.4) (2016-07-14)



<a name="2.0.3"></a>
## [2.0.3](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v2.0.2...v2.0.3) (2016-05-30)



<a name="2.0.2"></a>
## [2.0.2](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v2.0.1...v2.0.2) (2016-01-25)




<a name="2.0.1"></a>
## [2.0.1](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v2.0.0...v2.0.1) (2016-01-22)




<a name="2.0.0"></a>
# [2.0.0](https://github.com/tomchentw/unused-files-webpack-plugin/compare/v1.3.0...v2.0.0) (2016-01-22)


### Features

* **src:** rewrite in ES2015 format ([9a61f21](https://github.com/tomchentw/unused-files-webpack-plugin/commit/9a61f21))


### BREAKING CHANGES

* src: Removes commonjs module support.

Before:

```js
// webpack.config.js
var UnusedFilesWebpackPlugin = require("unused-files-webpack-plugin");
```

After:

In ES2015 module format:

```js
import UnusedFilesWebpackPlugin from "unused-files-webpack-plugin";
// it's the same as
import { default as UnusedFilesWebpackPlugin } from "unused-files-webpack-plugin";
// You could access from named export as well.
import { UnusedFilesWebpackPlugin } from "unused-files-webpack-plugin";
```

If you still use commonjs:

```js
var UnusedFilesWebpackPlugin = require("unused-files-webpack-plugin").default;
// or named export
var UnusedFilesWebpackPlugin = require("unused-files-webpack-plugin").UnusedFilesWebpackPlugin;
// with destructive assignment
var { UnusedFilesWebpackPlugin } = require("unused-files-webpack-plugin");
```



<a name"1.3.0"></a>
## 1.3.0 (2015-09-07)


#### Features


lib/__tests__/index-test.js
"use strict";

var Path = require("path");
var test = require("tape");
var MemoryFS = require("memory-fs");
var webpack = require("webpack");

var UnusedFilesWebpackPlugin = require("../index");

var EXPECTED_FILENAME_LIST = [
  "CHANGELOG.md",
  "README.md",
  "lib/__tests__/index-test.js",
  "package.json",
];

test("UnusedFilesWebpackPlugin", function (t) {
  var compiler = webpack({
    context: Path.resolve(__dirname, "../../"),
    entry: {
      UnusedFilesWebpackPlugin: Path.resolve(__dirname, "../index.js"),
    },
    output: {
      path: __dirname, // It will be in MemoryFS :)
    },
    plugins: [new UnusedFilesWebpackPlugin()],
  });
  compiler.outputFileSystem = new MemoryFS();

  compiler.run(function (err, stats) {
    t.equal(err, null);

    var warnings = stats.compilation.warnings;
    t.equal(warnings.length, 1);

    var unusedFilesError = warnings[0];
    t.equal(unusedFilesError instanceof Error, true);

    var message = unusedFilesError.message;
    var containsExpected = EXPECTED_FILENAME_LIST.every(function (filename) {
      return message.match(filename);
    });
    t.equal(containsExpected, true);

    t.end();
  });
});


lib/__tests__/index.spec.js
var gte = require("semver").gte;

var version = process.version;

if (gte(version, "8.0.0")) {
  module.exports = require("./index__8.0.0__.spec.js");
} else if (gte(version, "6.0.0")) {
  module.exports = require("./index__6.0.0__.spec.js");
} else if (gte(version, "4.0.0")) {
  module.exports = require("./index__4.0.0__.spec.js");
} else if (gte(version, "0.12.0")) {
  module.exports = require("./index__0.12.0__.spec.js");
} else {
  module.exports = require("./index__0.10.0__.spec.js");
}

src/__tests__/index.spec.js
import path from "path";
import MemoryFS from "memory-fs";
import webpack from "webpack";
import UnusedFilesWebpackPlugin from "../index";

const EXPECTED_FILENAME_LIST = [
  `CHANGELOG.md`,
  `README.md`,
  `src/__tests__/index.spec.js`
];

describe(`UnusedFilesWebpackPlugin module`, () => {
  it(
    `should work as expected`,
    done => {
      const compiler = webpack({
        context: path.resolve(__dirname, `../../`),
        entry: {
          UnusedFilesWebpackPlugin: path.resolve(__dirname, `../index.js`)
        },
        mode: "development",
        output: {
          path: __dirname // It will be in MemoryFS :)
        },
        plugins: [new UnusedFilesWebpackPlugin()],
        resolve: { fallback: { path: false } }
      });
      compiler.outputFileSystem = new MemoryFS();

      compiler.run((err, stats) => {
        expect(err).toBeFalsy();

        const { warnings } = stats.compilation;
        expect(warnings).toHaveLength(1);

        const [unusedFilesError] = warnings;
        expect(unusedFilesError).toBeInstanceOf(Error);

        const { message } = unusedFilesError;
        const containsExpected = EXPECTED_FILENAME_LIST.every(filename =>
          message.match(filename)
        );
        expect(containsExpected).toBeTruthy();

        done();
      });
    },
    10000
  );

  describe(`deprecated options.pattern`, () => {
    it(
      `should work as expected`,
      done => {
        const compiler = webpack({
          context: path.resolve(__dirname, `../../`),
          entry: {
            UnusedFilesWebpackPlugin: path.resolve(__dirname, `../index.js`)
          },
          mode: "development",
          output: {
            path: __dirname
          },
          plugins: [
            new UnusedFilesWebpackPlugin({
              pattern: "src/**/*.*"
            })
          ],
          resolve: { fallback: { path: false } }
        });
        compiler.outputFileSystem = new MemoryFS();

        compiler.run((err, stats) => {
          expect(err).toBeFalsy();

          const { warnings } = stats.compilation;
          expect(warnings).toHaveLength(1);

          const [unusedFilesError] = warnings;
          expect(unusedFilesError).toBeInstanceOf(Error);

          const { message } = unusedFilesError;
          expect(message).toMatchSnapshot();

          done();
        });
      },
      10000
    );
  });

  describe(`options.patterns`, () => {
    it(
      `should work as expected for a single pattern`,
      done => {
        const compiler = webpack({
          context: path.resolve(__dirname, `../../`),
          entry: {
            UnusedFilesWebpackPlugin: path.resolve(__dirname, `../index.js`)
          },
          mode: "development",
          output: {
            path: __dirname
          },
          plugins: [
            new UnusedFilesWebpackPlugin({
              patterns: ["src/**/*.*"]
            })
          ],
          resolve: { fallback: { path: false } }
        });
        compiler.outputFileSystem = new MemoryFS();

        compiler.run((err, stats) => {
          expect(err).toBeFalsy();

          const { warnings } = stats.compilation;
          expect(warnings).toHaveLength(1);

          const [unusedFilesError] = warnings;
          expect(unusedFilesError).toBeInstanceOf(Error);

          const { message } = unusedFilesError;
          expect(message).toMatchSnapshot();

          done();
        });
      },
      10000
    );

    it(
      `should work as expected for an array of patterns`,
      done => {
        const compiler = webpack({
          context: path.resolve(__dirname, `../../`),
          entry: {
            UnusedFilesWebpackPlugin: path.resolve(__dirname, `../index.js`)
          },
          mode: "development",
          output: {
            path: __dirname
          },
          plugins: [
            new UnusedFilesWebpackPlugin({
              patterns: ["src/**/*.*", "!**/__snapshots__/**/*.*"]
            })
          ],
          resolve: { fallback: { path: false } }
        });
        compiler.outputFileSystem = new MemoryFS();

        compiler.run((err, stats) => {
          expect(err).toBeFalsy();

          const { warnings } = stats.compilation;
          expect(warnings).toHaveLength(1);

          const [unusedFilesError] = warnings;
          expect(unusedFilesError).toBeInstanceOf(Error);

          const { message } = unusedFilesError;
          expect(message).toMatchSnapshot();

          done();
        });
      },
      10000
    );
  });
});


src/index.js
import path from "path";
import warning from "warning";
import nativeGlobAll from "glob-all";
import promisify from "util.promisify";

const globAll = promisify(nativeGlobAll);

function globOptionsWith(compiler, globOptions) {
  return {
    cwd: compiler.context,
    ...globOptions
  };
}

function getFileDepsMap(compilation) {
  const fileDepsBy = [...compilation.fileDependencies].reduce(
    (acc, usedFilepath) => {
      acc[usedFilepath] = true;
      return acc;
    },
    {}
  );

  const { assets } = compilation;
  Object.keys(assets).forEach(assetRelpath => {
    const existsAt = assets[assetRelpath].existsAt;
    fileDepsBy[existsAt] = true;
  });
  return fileDepsBy;
}

async function applyAfterEmit(compiler, compilation, plugin) {
  try {
    const globOptions = globOptionsWith(compiler, plugin.globOptions);
    const fileDepsMap = getFileDepsMap(compilation);

    const files = await globAll(
      plugin.options.patterns || plugin.options.pattern,
      globOptions
    );
    const unused = files.filter(
      it => !fileDepsMap[path.join(globOptions.cwd, it)]
    );

    if (unused.length !== 0) {
      throw new Error(`
UnusedFilesWebpackPlugin found some unused files:
${unused.join(`\n`)}`);
    }
  } catch (error) {
    if (plugin.options.failOnUnused && compilation.bail) {
      throw error;
    }
    const errorsList = plugin.options.failOnUnused
      ? compilation.errors
      : compilation.warnings;
    errorsList.push(error);
  }
}

export class UnusedFilesWebpackPlugin {
  constructor(options = {}) {
    warning(
      !options.pattern,
      `
"options.pattern" is deprecated and will be removed in v4.0.0.
Use "options.patterns" instead, which supports array of patterns and exclude pattern.
See https://www.npmjs.com/package/glob-all#notes
`
    );
    this.options = {
      ...options,
      patterns: options.patterns || options.pattern || [`**/*.*`],
      failOnUnused: options.failOnUnused === true
    };

    this.globOptions = {
      ignore: `node_modules/**/*`,
      ...options.globOptions
    };
  }

  apply(compiler) {
    compiler.hooks.afterEmit.tapAsync(
      UnusedFilesWebpackPlugin.name,
      (compilation, done) => {
        applyAfterEmit(compiler, compilation, this).then(done, done);
      }
    );
  }
}

export default UnusedFilesWebpackPlugin;


lib/index__4.0.0__.js
"use strict";

var _interopRequireDefault = require("babel-runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.UnusedFilesWebpackPlugin = void 0;

var _classCallCheck2 = _interopRequireDefault(
  require("babel-runtime/helpers/classCallCheck")
);

var _createClass2 = _interopRequireDefault(
  require("babel-runtime/helpers/createClass")
);

var _regenerator = _interopRequireDefault(require("babel-runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(
  require("babel-runtime/helpers/asyncToGenerator")
);

var _keys = _interopRequireDefault(
  require("babel-runtime/core-js/object/keys")
);

var _toConsumableArray2 = _interopRequireDefault(
  require("babel-runtime/helpers/toConsumableArray")
);

var _extends2 = _interopRequireDefault(
  require("babel-runtime/helpers/extends")
);

var _path = _interopRequireDefault(require("path"));

var _warning = _interopRequireDefault(require("warning"));

var _globAll = _interopRequireDefault(require("glob-all"));

var _util = _interopRequireDefault(require("util.promisify"));

var applyAfterEmit = (function() {
  var _ref = (0, _asyncToGenerator2.default)(
    /*#__PURE__*/
    _regenerator.default.mark(function _callee(compiler, compilation, plugin) {
      var globOptions, fileDepsMap, files, unused, errorsList;
      return _regenerator.default.wrap(
        function _callee$(_context) {
          while (1) {
            switch ((_context.prev = _context.next)) {
              case 0:
                _context.prev = 0;
                globOptions = globOptionsWith(compiler, plugin.globOptions);
                fileDepsMap = getFileDepsMap(compilation);
                _context.next = 5;
                return globAll(
                  plugin.options.patterns || plugin.options.pattern,
                  globOptions
                );

              case 5:
                files = _context.sent;
                unused = files.filter(function(it) {
                  return !fileDepsMap[_path.default.join(globOptions.cwd, it)];
                });

                if (!(unused.length !== 0)) {
                  _context.next = 9;
                  break;
                }

                throw new Error(`
UnusedFilesWebpackPlugin found some unused files:
${unused.join(`\n`)}`);

              case 9:
                _context.next = 17;
                break;

              case 11:
                _context.prev = 11;
                _context.t0 = _context["catch"](0);

                if (!(plugin.options.failOnUnused && compilation.bail)) {
                  _context.next = 15;
                  break;
                }

                throw _context.t0;

              case 15:
                errorsList = plugin.options.failOnUnused
                  ? compilation.errors
                  : compilation.warnings;
                errorsList.push(_context.t0);

              case 17:
              case "end":
                return _context.stop();
            }
          }
        },
        _callee,
        this,
        [[0, 11]]
      );
    })
  );

  return function applyAfterEmit(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
})();

var globAll = (0, _util.default)(_globAll.default);

function globOptionsWith(compiler, globOptions) {
  return (0, _extends2.default)(
    {
      cwd: compiler.context
    },
    globOptions
  );
}

function getFileDepsMap(compilation) {
  var fileDepsBy = []
    .concat((0, _toConsumableArray2.default)(compilation.fileDependencies))
    .reduce(function(acc, usedFilepath) {
      acc[usedFilepath] = true;
      return acc;
    }, {});
  var assets = compilation.assets;
  (0, _keys.default)(assets).forEach(function(assetRelpath) {
    var existsAt = assets[assetRelpath].existsAt;
    fileDepsBy[existsAt] = true;
  });
  return fileDepsBy;
}

var UnusedFilesWebpackPlugin =
  /*#__PURE__*/
  (function() {
    function UnusedFilesWebpackPlugin() {
      var options =
        arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      (0, _classCallCheck2.default)(this, UnusedFilesWebpackPlugin);
      (0, _warning.default)(
        !options.pattern,
        `
"options.pattern" is deprecated and will be removed in v4.0.0.
Use "options.patterns" instead, which supports array of patterns and exclude pattern.
See https://www.npmjs.com/package/glob-all#notes
`
      );
      this.options = (0, _extends2.default)({}, options, {
        patterns: options.patterns || options.pattern || [`**/*.*`],
        failOnUnused: options.failOnUnused === true
      });
      this.globOptions = (0, _extends2.default)(
        {
          ignore: `node_modules/**/*`
        },
        options.globOptions
      );
    }

    (0, _createClass2.default)(UnusedFilesWebpackPlugin, [
      {
        key: "apply",
        value: function apply(compiler) {
          var _this = this;

          compiler.hooks.afterEmit.tapAsync(
            UnusedFilesWebpackPlugin.name,
            function(compilation, done) {
              applyAfterEmit(compiler, compilation, _this).then(done, done);
            }
          );
        }
      }
    ]);
    return UnusedFilesWebpackPlugin;
  })();

exports.UnusedFilesWebpackPlugin = UnusedFilesWebpackPlugin;
var _default = UnusedFilesWebpackPlugin;
exports.default = _default;


lib/index__6.0.0__.js
"use strict";

var _interopRequireDefault = require("babel-runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.UnusedFilesWebpackPlugin = void 0;

var _asyncToGenerator2 = _interopRequireDefault(
  require("babel-runtime/helpers/asyncToGenerator")
);

var _keys = _interopRequireDefault(
  require("babel-runtime/core-js/object/keys")
);

var _extends2 = _interopRequireDefault(
  require("babel-runtime/helpers/extends")
);

var _path = _interopRequireDefault(require("path"));

var _warning = _interopRequireDefault(require("warning"));

var _globAll = _interopRequireDefault(require("glob-all"));

var _util = _interopRequireDefault(require("util.promisify"));

let applyAfterEmit = (() => {
  var _ref = (0, _asyncToGenerator2.default)(function*(
    compiler,
    compilation,
    plugin
  ) {
    try {
      const globOptions = globOptionsWith(compiler, plugin.globOptions);
      const fileDepsMap = getFileDepsMap(compilation);
      const files = yield globAll(
        plugin.options.patterns || plugin.options.pattern,
        globOptions
      );
      const unused = files.filter(
        it => !fileDepsMap[_path.default.join(globOptions.cwd, it)]
      );

      if (unused.length !== 0) {
        throw new Error(`
UnusedFilesWebpackPlugin found some unused files:
${unused.join(`\n`)}`);
      }
    } catch (error) {
      if (plugin.options.failOnUnused && compilation.bail) {
        throw error;
      }

      const errorsList = plugin.options.failOnUnused
        ? compilation.errors
        : compilation.warnings;
      errorsList.push(error);
    }
  });

  return function applyAfterEmit(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
})();

const globAll = (0, _util.default)(_globAll.default);

function globOptionsWith(compiler, globOptions) {
  return (0, _extends2.default)(
    {
      cwd: compiler.context
    },
    globOptions
  );
}

function getFileDepsMap(compilation) {
  const fileDepsBy = [...compilation.fileDependencies].reduce(
    (acc, usedFilepath) => {
      acc[usedFilepath] = true;
      return acc;
    },
    {}
  );
  const assets = compilation.assets;
  (0, _keys.default)(assets).forEach(assetRelpath => {
    const existsAt = assets[assetRelpath].existsAt;
    fileDepsBy[existsAt] = true;
  });
  return fileDepsBy;
}

class UnusedFilesWebpackPlugin {
  constructor(options = {}) {
    (0, _warning.default)(
      !options.pattern,
      `
"options.pattern" is deprecated and will be removed in v4.0.0.
Use "options.patterns" instead, which supports array of patterns and exclude pattern.
See https://www.npmjs.com/package/glob-all#notes
`
    );
    this.options = (0, _extends2.default)({}, options, {
      patterns: options.patterns || options.pattern || [`**/*.*`],
      failOnUnused: options.failOnUnused === true
    });
    this.globOptions = (0, _extends2.default)(
      {
        ignore: `node_modules/**/*`
      },
      options.globOptions
    );
  }

  apply(compiler) {
    compiler.hooks.afterEmit.tapAsync(
      UnusedFilesWebpackPlugin.name,
      (compilation, done) => {
        applyAfterEmit(compiler, compilation, this).then(done, done);
      }
    );
  }
}

exports.UnusedFilesWebpackPlugin = UnusedFilesWebpackPlugin;
var _default = UnusedFilesWebpackPlugin;
exports.default = _default;


lib/index__8.0.0__.js
"use strict";

var _interopRequireDefault = require("babel-runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.UnusedFilesWebpackPlugin = void 0;

var _keys = _interopRequireDefault(
  require("babel-runtime/core-js/object/keys")
);

var _extends2 = _interopRequireDefault(
  require("babel-runtime/helpers/extends")
);

var _path = _interopRequireDefault(require("path"));

var _warning = _interopRequireDefault(require("warning"));

var _globAll = _interopRequireDefault(require("glob-all"));

var _util = _interopRequireDefault(require("util.promisify"));

const globAll = (0, _util.default)(_globAll.default);

function globOptionsWith(compiler, globOptions) {
  return (0, _extends2.default)(
    {
      cwd: compiler.context
    },
    globOptions
  );
}

function getFileDepsMap(compilation) {
  const fileDepsBy = [...compilation.fileDependencies].reduce(
    (acc, usedFilepath) => {
      acc[usedFilepath] = true;
      return acc;
    },
    {}
  );
  const { assets } = compilation;
  (0, _keys.default)(assets).forEach(assetRelpath => {
    const existsAt = assets[assetRelpath].existsAt;
    fileDepsBy[existsAt] = true;
  });
  return fileDepsBy;
}

async function applyAfterEmit(compiler, compilation, plugin) {
  try {
    const globOptions = globOptionsWith(compiler, plugin.globOptions);
    const fileDepsMap = getFileDepsMap(compilation);
    const files = await globAll(
      plugin.options.patterns || plugin.options.pattern,
      globOptions
    );
    const unused = files.filter(
      it => !fileDepsMap[_path.default.join(globOptions.cwd, it)]
    );

    if (unused.length !== 0) {
      throw new Error(`
UnusedFilesWebpackPlugin found some unused files:
${unused.join(`\n`)}`);
    }
  } catch (error) {
    if (plugin.options.failOnUnused && compilation.bail) {
      throw error;
    }

    const errorsList = plugin.options.failOnUnused
      ? compilation.errors
      : compilation.warnings;
    errorsList.push(error);
  }
}

class UnusedFilesWebpackPlugin {
  constructor(options = {}) {
    (0, _warning.default)(
      !options.pattern,
      `
"options.pattern" is deprecated and will be removed in v4.0.0.
Use "options.patterns" instead, which supports array of patterns and exclude pattern.
See https://www.npmjs.com/package/glob-all#notes
`
    );
    this.options = (0, _extends2.default)({}, options, {
      patterns: options.patterns || options.pattern || [`**/*.*`],
      failOnUnused: options.failOnUnused === true
    });
    this.globOptions = (0, _extends2.default)(
      {
        ignore: `node_modules/**/*`
      },
      options.globOptions
    );
  }

  apply(compiler) {
    compiler.hooks.afterEmit.tapAsync(
      UnusedFilesWebpackPlugin.name,
      (compilation, done) => {
        applyAfterEmit(compiler, compilation, this).then(done, done);
      }
    );
  }
}

exports.UnusedFilesWebpackPlugin = UnusedFilesWebpackPlugin;
var _default = UnusedFilesWebpackPlugin;
exports.default = _default;


lib/index__0.10.0__.js
"use strict";

var _interopRequireDefault = require("babel-runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.UnusedFilesWebpackPlugin = void 0;

var _classCallCheck2 = _interopRequireDefault(
  require("babel-runtime/helpers/classCallCheck")
);

var _createClass2 = _interopRequireDefault(
  require("babel-runtime/helpers/createClass")
);

var _regenerator = _interopRequireDefault(require("babel-runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(
  require("babel-runtime/helpers/asyncToGenerator")
);

var _keys = _interopRequireDefault(
  require("babel-runtime/core-js/object/keys")
);

var _toConsumableArray2 = _interopRequireDefault(
  require("babel-runtime/helpers/toConsumableArray")
);

var _extends2 = _interopRequireDefault(
  require("babel-runtime/helpers/extends")
);

var _path = _interopRequireDefault(require("path"));

var _warning = _interopRequireDefault(require("warning"));

var _globAll = _interopRequireDefault(require("glob-all"));

var _util = _interopRequireDefault(require("util.promisify"));

var applyAfterEmit = (function() {
  var _ref = (0, _asyncToGenerator2.default)(
    /*#__PURE__*/
    _regenerator.default.mark(function _callee(compiler, compilation, plugin) {
      var globOptions, fileDepsMap, files, unused, errorsList;
      return _regenerator.default.wrap(
        function _callee$(_context) {
          while (1) {
            switch ((_context.prev = _context.next)) {
              case 0:
                _context.prev = 0;
                globOptions = globOptionsWith(compiler, plugin.globOptions);
                fileDepsMap = getFileDepsMap(compilation);
                _context.next = 5;
                return globAll(
                  plugin.options.patterns || plugin.options.pattern,
                  globOptions
                );

              case 5:
                files = _context.sent;
                unused = files.filter(function(it) {
                  return !fileDepsMap[_path.default.join(globOptions.cwd, it)];
                });

                if (!(unused.length !== 0)) {
                  _context.next = 9;
                  break;
                }

                throw new Error(
                  "\nUnusedFilesWebpackPlugin found some unused files:\n".concat(
                    unused.join("\n")
                  )
                );

              case 9:
                _context.next = 17;
                break;

              case 11:
                _context.prev = 11;
                _context.t0 = _context["catch"](0);

                if (!(plugin.options.failOnUnused && compilation.bail)) {
                  _context.next = 15;
                  break;
                }

                throw _context.t0;

              case 15:
                errorsList = plugin.options.failOnUnused
                  ? compilation.errors
                  : compilation.warnings;
                errorsList.push(_context.t0);

              case 17:
              case "end":
                return _context.stop();
            }
          }
        },
        _callee,
        this,
        [[0, 11]]
      );
    })
  );

  return function applyAfterEmit(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
})();

var globAll = (0, _util.default)(_globAll.default);

function globOptionsWith(compiler, globOptions) {
  return (0, _extends2.default)(
    {
      cwd: compiler.context
    },
    globOptions
  );
}

function getFileDepsMap(compilation) {
  var fileDepsBy = []
    .concat((0, _toConsumableArray2.default)(compilation.fileDependencies))
    .reduce(function(acc, usedFilepath) {
      acc[usedFilepath] = true;
      return acc;
    }, {});
  var assets = compilation.assets;
  (0, _keys.default)(assets).forEach(function(assetRelpath) {
    var existsAt = assets[assetRelpath].existsAt;
    fileDepsBy[existsAt] = true;
  });
  return fileDepsBy;
}

var UnusedFilesWebpackPlugin =
  /*#__PURE__*/
  (function() {
    function UnusedFilesWebpackPlugin() {
      var options =
        arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      (0, _classCallCheck2.default)(this, UnusedFilesWebpackPlugin);
      (0, _warning.default)(
        !options.pattern,
        '\n"options.pattern" is deprecated and will be removed in v4.0.0.\nUse "options.patterns" instead, which supports array of patterns and exclude pattern.\nSee https://www.npmjs.com/package/glob-all#notes\n'
      );
      this.options = (0, _extends2.default)({}, options, {
        patterns: options.patterns || options.pattern || ["**/*.*"],
        failOnUnused: options.failOnUnused === true
      });
      this.globOptions = (0, _extends2.default)(
        {
          ignore: "node_modules/**/*"
        },
        options.globOptions
      );
    }

    (0, _createClass2.default)(UnusedFilesWebpackPlugin, [
      {
        key: "apply",
        value: function apply(compiler) {
          var _this = this;

          compiler.hooks.afterEmit.tapAsync(
            UnusedFilesWebpackPlugin.name,
            function(compilation, done) {
              applyAfterEmit(compiler, compilation, _this).then(done, done);
            }
          );
        }
      }
    ]);
    return UnusedFilesWebpackPlugin;
  })();

exports.UnusedFilesWebpackPlugin = UnusedFilesWebpackPlugin;
var _default = UnusedFilesWebpackPlugin;
exports.default = _default;
